api/app/registration/api.py

```diff
@@ -1,21 +1,191 @@
+    }
+
+
+def offer_update_info(offer_entity):
+    return {
+        'user_id': offer_entity.user_id,
+        'event_id': offer_entity.event_id,
+        'offer_date': offer_entity.offer_date,
+        'expiry_date': offer_entity.expiry_date,
+        'payment_required': offer_entity.payment_required,
+        'travel_award': offer_entity.travel_award,
+        'accommodation_award': offer_entity.accommodation_award,
+        'rejected': offer_entity.rejected,
+        'accepted': offer_entity.accepted,
+        'rejected_reason': offer_entity.rejected_reason,
+        'updated_at': offer_entity.updated_at,
+    }
+
+
+class OfferAPI(OfferMixin, restful.Resource):
+    offer_fields = {
+        'user_id': fields.Integer,
+        'event_id': fields.Integer,
+        'offer_date': fields.DateTime,
+        'expiry_date': fields.DateTime,
+        'payment_required': fields.Boolean,
+        'travel_award': fields.Boolean,
+        'accommodation_award': fields.Boolean,
+        'rejected': fields.Boolean,
+        'accepted': fields.Boolean,
+        'rejected_reason': fields.String,
+        'updated_at': fields.DateTime,
+
+
+    }
+
+    @auth_required
+    @marshal_with(offer_fields)
+    def put(self):
+        # update existing offer
+        args = self.req_parser.parse_args()
+        offer_id = args['offer_id']
+        accepted = args['accepted']
+        rejected = args['rejected']
+        rejected_reason = args['rejected_reason']
+        offer = db.session.query(Offer).filter(Offer.id == offer_id).first()
+
+        if not offer:
+            return errors.OFFER_NOT_FOUND
+
+        try:
+            user_id = verify_token(request.headers.get('Authorization'))['id']
+
+            if offer and (not offer.user_id == user_id):
+                return errors.FORBIDDEN
+
+            offer.updated_at = datetime.now()
+            offer.accepted = accepted
+            offer.rejected = rejected
+            offer.rejected_reason = rejected_reason
+
+            db.session.commit()
+
+        except:
+            LOGGER.error("Failed to update offer with id {}".format(args['offer_id']))
+            return errors.ADD_OFFER_FAILED
+        return offer_update_info(offer), 201
+
+    @admin_required
+    @marshal_with(offer_fields)
+    def post(self):
+        args = self.req_parser.parse_args()
+        user_id = args['user_id']
+        email = g.current_user['email']
+        event_id = args['event_id']
+        offer_date = datetime.strptime((args['offer_date']), '%Y-%m-%dT%H:%M:%S.%fZ')
+        expiry_date = datetime.strptime((args['expiry_date']), '%Y-%m-%dT%H:%M:%S.%fZ')
+        payment_required = args['payment_required']
+        travel_award = args['travel_award']
+        accommodation_award = args['accommodation_award']
+        updated_at = datetime.strptime((args['updated_at']), '%Y-%m-%dT%H:%M:%S.%fZ')
+        user = db.session.query(AppUser).filter(AppUser.email == email).first()
```

Comment:
This user object is used later for sending the offer email, but is based on g.current_user['email'] - which is actually the admin user making the offer, rather than the user receiving the offer. Please change this to lookup the user based on user_id